hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/TimelineService.java

```diff
@@ -344,25 +345,36 @@ private int startServiceOnPort(int port) throws IOException {
       // Returns port to try when trying to bind a service. Handles wrapping and skipping privileged ports.
       int tryPort = port == 0 ? port : (port + attempt - 1024) % (65536 - 1024) + 1024;
       try {
+        createApp();
         app.start(tryPort);
         return app.port();
       } catch (Exception e) {
-        if (e.getMessage() != null && e.getMessage().contains("Failed to bind to")) {
```

Comment:
Is this line an error since it is introduced, is there any behavior change for Javaline itself?

Reply:
This is just a better way to do error classification. If a developer in Javalin changes the string then your handling will break so that is why you typically catch a specific exception type instead of simply `Exception`. 

The bug is up at line 348. The underlying server object is the same and the code tries to call start on it twice and the underlying port does not change.

Reply:
> The bug is up at line 348

I didn't quite get it, the original logic invoke `startService` only once, where a singleton Javaline app would be created, and the `startServiceOnPort` would try to start a port on the app then returns, so the actual port would be updated.

Reply:
You can take the test case and run it against master to better see the issue. There is a server object that is passed in to Javalin that is not updated and will have the old port set.

Reply:
It looks like the server started by Javaline is kind of sinlgeton within one JVM process, and the app can not start twice if the server already got started on a port. Do we have a scenario that multiple timeline server got triggered to start in one JVM?

Reply:
This logic can be hit if the port is already in use on the machine. It doesn't necessarily need to be a timeline server. You can also hit this if you have multiple write clients on the same machine with timeline server enabled.

Reply:
> This logic can be hit if the port is already in use on the machine

Can you supplement the test case to cover this case?

Reply:
The test case is doing this currently, can you be more specific about what you're looking for?

Reply:
> This logic can be hit if the port is already in use on the machine

I'm not convinced by this, the used port is occupied by the Javaline app itselft, I'm wondering how it behaves if the port is assigned by stuff that is not Javaline.

Reply:
ok, I will add another case with some lightweight server option

Reply:
Added a new test